bindings/java/src/org/sleuthkit/datamodel/ScoringManager.java

```diff
@@ -174,22 +182,38 @@ private void setAggregateScore(long objId, long dataSourceObjectId, Score score,
 	 */
 	Score updateAggregateScore(long objId, long dataSourceObjectId, Score resultScore, CaseDbTransaction transaction) throws TskCoreException {
 
-		// Get the current score 
-		Score currentScore = ScoringManager.this.getAggregateScore(objId, transaction);
-
-		// If current score is Unknown And newscore is not Unknown - allow None (good) to be recorded
-		// or if the new score is higher than the current score
-		if  ( (currentScore.compareTo(Score.SCORE_UNKNOWN) == 0 && resultScore.compareTo(Score.SCORE_UNKNOWN) != 0)
-			  || (Score.getScoreComparator().compare(resultScore, currentScore) > 0)) {
-			ScoringManager.this.setAggregateScore(objId, dataSourceObjectId, resultScore, transaction);
-			
-			// register score change in the transaction.
-			transaction.registerScoreChange(new ScoreChange(objId, dataSourceObjectId, currentScore, resultScore));
-			return resultScore;
-		} else {
-			// return the current score
-			return currentScore;
+		// We use a version based optimistic row locking to prevent simultaneous updates to aggregate score. 
+		// If another thread interrupts the update, we get a SQLException and retry. 
+		int numRetries = 0;
+		while (++numRetries < 10) {
+			try {
+
+				// Get the current score and it's version
+				Pair<Score, Integer> currentScoreAndVersion = ScoringManager.this.getAggregateScoreAndVersion(objId, transaction.getConnection());
+
+				Score currentScore = currentScoreAndVersion.getKey();
+				Integer currentScoreVersion = currentScoreAndVersion.getRight();
+
+				// If current score is Unknown And newscore is not Unknown - allow None (good) to be recorded
+				// or if the new score is higher than the current score
+				if ((currentScore.compareTo(Score.SCORE_UNKNOWN) == 0 && resultScore.compareTo(Score.SCORE_UNKNOWN) != 0)
+						|| (Score.getScoreComparator().compare(resultScore, currentScore) > 0)) {
+
+					ScoringManager.this.setAggregateScore(objId, dataSourceObjectId, resultScore, currentScoreVersion, transaction);
+
+					// register score change in the transaction.
+					transaction.registerScoreChange(new ScoreChange(objId, dataSourceObjectId, currentScore, resultScore));
+					return resultScore;
+				} else {
+					// return the current score
+					return currentScore;
+				}
+			} catch (SQLException ex) {
+				LOGGER.log(Level.WARNING, String.format("Error trying to update aggregate score for objId = %d, data source object id = %d.  Will try again. ", objId, dataSourceObjectId), ex);
```

Comment:
You probably need to make a checkpoint and rollback the transaction here. I ran into errors with postgresql when I tried to continue using a transaction after an error. See https://www.progress.com/blogs/solution-to-current-transaction-is-aborted-error-for-postgresql-compatibility-with-odbc-jdbc

Reply:
Hmm.  The transaction is given to us by the calling client.  I don't think we can/should just roll it back  - it probably creates more trouble for the client than what we are trying to address.

@bcarrier, your thoughts?

Reply:
Ahh, I see what you mean - a savepoint for the transaction.  Yea - let me look into it.

Reply:
I did it in createHost() in HostManager.java